akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,433 @@
+
+  override def spawnAdapter[U](f: U ⇒ T): ActorRef[U] = {
+    val name = Helpers.base64(nextName)
+    nextName += 1
+    val ref = new FunctionRef[U](self.path / name,
+      (msg, _) ⇒ send(f(msg)),
+      (self) ⇒ sendSystem(DeathWatchNotification(self, null)))
+    childrenMap = childrenMap.updated(name, ref)
+    ref
+  }
+
+  private[this] var receiveTimeout: (FiniteDuration, T) = null
+  override def setReceiveTimeout(d: FiniteDuration, msg: T): Unit = {
+    if (Debug) println(s"$self setting receive timeout of $d, msg $msg")
+    receiveTimeout = (d, msg)
+  }
+  override def cancelReceiveTimeout(): Unit = {
+    if (Debug) println(s"$self canceling receive timeout")
+    receiveTimeout = null
+  }
+
+  /*
+   * Implementation of the invocation mechanics.
+   */
+
+  // see comment in companion object for details
+  @volatile private[this] var _status: Int = 0
+  protected[typed] def getStatus: Int = _status
+  private[this] val queue: Queue[T] = new ConcurrentLinkedQueue[T]
+  private[typed] def peekMessage: T = queue.peek()
+  private[this] val maxQueue: Int = Math.min(props.queueSize, maxActivations)
+  @volatile private[this] var _systemQueue: LatestFirstSystemMessageList = SystemMessageList.LNil
+
+  protected def maySend: Boolean = !isTerminating
+  protected def isTerminating: Boolean = ActorCell.isTerminating(_status)
+  protected def setTerminating(): Unit = if (!ActorCell.isTerminating(_status)) unsafe.getAndAddInt(this, statusOffset, terminatingBit)
+  protected def setClosed(): Unit = if (!isClosed(_status)) unsafe.getAndAddInt(this, statusOffset, closedBit)
+
+  private def handleException: Catcher[Unit] = {
+    case e: InterruptedException ⇒
+      publish(Error(e, self.path.toString, getClass, "interrupted during message send"))
+      Thread.currentThread.interrupt()
+    case NonFatal(e) ⇒
+      publish(Error(e, self.path.toString, getClass, "swallowing exception during message send"))
+  }
+
+  def send(msg: T): Unit =
+    try {
+      val old = unsafe.getAndAddInt(this, statusOffset, 1)
+      val oldActivations = activations(old)
+      // this is not an off-by-one: #msgs is activations-1 if >0
+      if (oldActivations > maxQueue) {
+        if (Debug) println(s"[$thread] $self NOT enqueueing $msg at status $old ($oldActivations > $maxQueue)")
+        // cannot enqueue, need to give back activation token
+        unsafe.getAndAddInt(this, statusOffset, -1)
+        system.eventStream.publish(Dropped(msg, self))
+      } else if (ActorCell.isTerminating(old)) {
+        if (Debug) println(s"[$thread] $self NOT enqueueing $msg at status $old (is terminating)")
+        unsafe.getAndAddInt(this, statusOffset, -1)
+        system.deadLetters ! msg
+      } else {
+        if (Debug) println(s"[$thread] $self enqueueing $msg at status $old")
+        // need to enqueue; if the actor sees the token but not the message, it will reschedule
```

Comment:
Aha! So this explains the looking at the number of activations in `run` - so if we processed less, because we didn't see the message there then we reschedule?